Add database

JRRT93 opened the pull request:
Добавлена БД, новые имплементации хранилищ, новые эндпоинты и тесты согласно ТЗ

JRRT93 committed:
feat: ERD в README

JRRT93 committed:
feat: подключена БД H2, реализована новая бизнес-логика по получению рейтинга и жанров согласно ТЗ, добавлены новые тесты

JRRT93 committed:
fix: поправлено форматирования строк после фейла теста чек-стайла

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDaoImpl.java:
```diff
@@ -0,0 +1,148 @@
+            .releaseDate(rs.getDate("release_date").toLocalDate())
+            .likes(null)
+            .mpa(null)
+            .genres(null)
```
Когда заполняете поле, которое имеет тип "Коллекция" лучше передавать пустую коллекцию, а не null

JRRT93 replied:
Исправлено

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDaoImpl.java:
```diff
@@ -0,0 +1,148 @@
+    public Film findFilmById(long id) throws ValidationException {
+        String sqlQuery = "SELECT * FROM films WHERE film_id = ?;";
+        Film film;
+        try {
```
У вас в целом норм, но best practice - использовать Optional
```
@Override
public Film findFilmById(long id) throws ValidationException {
    String sqlQuery = "SELECT * FROM films WHERE film_id = ?;";
    Film film = Optional.ofNullable(jdbcTemplate.queryForObject(sqlQuery, rowMapper, id))
            .orElseThrow(() -> new ValidationException("Incorrect ID=" + id + ". This film is not in database yet"));
    film.setGenres(new TreeSet<>(extractGenres(film)));
    film.setLikes(new HashSet<>(extractLikes(film)));
    film.setMpa(extractRating(film));
    return film;
}
```

JRRT93 replied:
Изменил методы findById() для всех сущностей. Но вышла накладка - я попытался использовать конструкцию Optional.ofNullable().elseThrow() прям в DAO, но jdbcTemplate.queryForObject не возвращает null и просто выбрасывает ошибку EmptyResultDataAccessException при отсутствии записи в БД и код 500 от сервера, а это нас не устраивает, потому что в тестах требуется, чтобы выбрасывалось кастомное исключение и код ответ 404. Поэтому сейчас реализовано так - метод через конструкцию try-catch пытается сделать сущность и вернуть её как Optional.of() или если была поймана EmptyResultDataAccessException  возвращает Optional.empty(). А уже там где вызывался метод DAO findById() проводится проверка и через .elseThrow()  выбрасывается кастомное исключение

VladimirTitov4 replied:
Понял, хорошо

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDaoImpl.java:
```diff
@@ -0,0 +1,148 @@
+
+    private List<Genre> extractGenres(Film film) {
+        String sqlQuery = "SELECT  f.genre_id, g.name\n" +
+                "FROM (SELECT genre_id FROM film_genre WHERE film_id = ?) as f\n" +
```
Попробуйте не использовать подзапрос, а использовать join

JRRT93 replied:
Исправлено, заменил подзапросы объединениями таблиц

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDaoImpl.java:
```diff
@@ -0,0 +1,148 @@
+
+    private Rating extractRating(Film film) {
+        String sqlQuery = "SELECT rating_id, name FROM ratings " +
+                "WHERE rating_id = (SELECT rating_mpa FROM films WHERE film_id = ?);";
```
Здесь и везде в остальных местах нужно заменить подзапрос на join

JRRT93 replied:
Исправлено, заменил подзапросы объединениями таблиц

VladimirTitov4 commented on src/main/java/ru/yandex/practicum/filmorate/dao/UserDaoImpl.java:
```diff
@@ -0,0 +1,97 @@
+        try {
+            user = jdbcTemplate.queryForObject(sqlQuery, rowMapper, id);
+        } catch (Exception e) {
+            user = null;
```
Здесь так же можно использовать мой предложенный вариант с Optional. В целом null возвращать из метода всегда не очень

JRRT93 replied:
Применил описанный выше подход для всех сущностей - Film, User, Rating, Genre

JRRT93 committed:
fix: в RowMapper'ы добавлена инициализация полей пустыми коллекциями, в SQL запросах подзапросы заменены объединениями таблиц
feat: для методов finById() изменен тип возвращаемого значения на Optional<T> и соответсвенно переделана реализация этих методов

JRRT93 committed:
fix: в RowMapper'ы добавлена инициализация полей пустыми коллекциями, в SQL запросах подзапросы заменены объединениями таблиц
feat: для методов finById() изменен тип возвращаемого значения на Optional<T> и соответсвенно переделана реализация этих методов

JRRT93 committed:
fix: checkstyle

JRRT93 committed:
Merge remote-tracking branch 'origin/add-database' into add-database
